### ISSUE TITLE
NPE when using import context.dispatcher

### ISSUE THREAD

**Issue by patriknw** [akka/akka#25040]
Reported here: https://groups.google.com/d/msg/akka-user/LHWnzwIWnOw/3hBGmncrBAAJ

E.g. if a future callback is running after the enclosing actor has been stopped.

It's because we try to clear things in the actor cell to avoid holding more memory references than necessary, in case the cell is still referenced (from ActorRef).

Can we clean more selectively to still keep the context? Otherwise we should revisit docs and don't use `import context.dispatcher` but rather `implicit val ec: ExecutionContext = context.dispatcher`.

**Comment by raboof** [comment]
I'd say showing `implicit val ec: ExecutionContext = context.dispatcher` is clearer/simpler, if a bit more verbose :+1:.

**Comment by johanandren** [comment]
I'd like to add to this that it is not in general unsafe to import the dispatcher like so, since the normal case is to call `schedule` or some transformation on `Future` as a reaction to a message - at that point the actor is still alive. 

The problem arises when something is closing over the actor context, and using the the dispatcher when scheduled later, so for example:

```
import context.dispatcher
def receive = {
 case message => 
    someFutureA.map { a => 
       // here we are implicitly referencing the dispatcher without being sure the actor is alive
       newFutureB.map(b => b)
    } // here it will always be safe to reference the dispatcher, since we couldn't act on a message if actor is not alive
```

I think this an anti pattern and I'm not convinced this is a reason to even change the docs to be honest.

**Comment by viktorklang** [comment]
For about a decade I've wanted to have a @CloseOverByValue annotation. Perhaps that's just a bandaid though.

**Comment by patriknw** [comment]
Someone hit this again. Even if we don't want to rewrite the internals we should remove `import context.dispatcher` from all documentation samples.

For example, this could fail:

```
class Foo(other: ActorRef) extends Actor {

  implicit val timout = Timeout(5.seconds)

  def receive = {
    case msg ⇒
      import context.dispatcher
      val reply = for {
        a ← other ? msg
        b ← other ? a
      } yield b
      reply.pipeTo(sender())
  }

}
```

Instead it should be like:

```
object Foo {
  def asyncInteraction(other: ActorRef, msg: Any)(implicit ec: ExecutionContext): Future[Result] = {
    implicit val timout = Timeout(5.seconds)
    val reply = for {
      a ← other ? msg
      b ← other ? a
    } yield Result(b)
  }
}

class Foo(other: ActorRef) extends Actor {

  def receive = {
    case msg ⇒
      implicit val ec = context.dispatcher
      val reply = Foo.asyncInteraction(other, msg)
      reply.pipeTo(sender())
  }

}
```

Note that in that case it is safe to use `import context.dispatcher` but we shouldn't show that.

**Comment by viktorklang** [comment]
One possible solution would be to deprecate `getDispatcher` and `dispatcher` on `ActorContext`, and instead introduce a `getExecutionContexg` and an `executionContext` method which is not implicit. And the deprecation message on `getDispatcher` and `dispatcher` could recommend to do `implicit val ec = context.executionContext` for Scala and `final ExecutionContext ec = context.getExecutionContext()` for Java`

**Comment by patriknw** [comment]
I agree, that also matches what we have in Typed.

**Comment by helena** [comment]
Is it incorrect in the first place to import the dispatcher there, if n-futures or even otherwise?
```scala
def receive = {
    case msg ⇒
      import context.dispatcher
      Future.Sequence(...)
  }
```

**Comment by jrudolph** [comment]
We could also just call it a bug, that Akka overwrites an immutable val and so circumvents Scala's "stable identifier" safety precautions for imports. Might make sense to see what breaks if we remove that code since that's also the single biggest performance problem of the Actor life-cycle.

**Comment by viktorklang** [comment]
It’s also a bug to close over mutable state beyond its lifecycle..
(ActorContext)
-- 
Cheers,
√

**Comment by patriknw** [comment]
I looked some more at the object references. The `_actor` field in the ActorCell is set to null when the actor is terminated, so I was wrong about that reference.

There shouldn't be any other references to the actor instance unless the application is referencing it and we can't do anything about that.

Even if the actor instance is referenced from somewhere I don't see that clearing the context would reduce the memory footprint of a stopped actor much since all things it references have a longer lifecycle.

<img width="1035" alt="screenshot 2019-02-18 at 13 45 24" src="https://user-images.githubusercontent.com/336161/52952932-5cf97f80-3386-11e9-8e0e-395482a33ce6.png">

The `props` could maybe close over some other things but `props` itself is changed to `terminatedProps` when the actor is stopped.

In other words, if we can't motivate why the `context` is cleared we should remove that.

**Comment by shkoder** [comment]
@patriknw should I resurrect #25332?

**Comment by patriknw** [comment]
@shkoder yes please that would be great

**Comment by renatocaval** [comment]
Is that still a `pick-next`? Less comment here is from 11/2019.

**Comment by patriknw** [comment]
yes, this should be completed. There is a promising draft PR but requires some cleanup and careful reviewing.

**Comment by patriknw** [comment]
Fixed in https://github.com/akka/akka/pull/27614
